stdlib/path.gr

```diff
@@ -0,0 +1,775 @@
+ * - Paths referencing files must not include trailing forward slashes, but paths referencing directories may
+ * - The path segment `.` indicates the relative "current" directory of a path, and `..` indicates the parent directory of a path
+ * 
+ * @example import Path from "path"
+ * 
+ * @since v0.5.5
+ */
+import String from "string"
+import List from "list"
+import Option from "option"
+import Result from "result"
+import Char from "char"
+
+// this module is influenced by https://github.com/reasonml/reason-native/blob/a0ddab6ab25237961e32d8732b0a222ec2372d4a/src/fp/Fp.re
+// with some modifications; reason-native license:
+
+// MIT License
+//
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+enum Token {
+  Slash,
+  Dot,
+  Dotdot,
+  DriveTok(Char),
+  Text(String),
+}
+
+enum DirsUp {
+  Zero,
+  Positive,
+}
+
+// Dummy record names put here just to distinguish the two. These could be
+// replaced with opaque types if they get added to the language
+record Relative {
+  _rel: Void,
+}
+
+record Absolute {
+  _abs: Void,
+}
+
+record File {
+  _file: Void,
+}
+
+record Directory {
+  _directory: Void,
+}
+
+enum FileType {
+  File,
+  Dir,
+}
+
+// hack to be able to concretely distinguish TypedPath from PathInfo and
+// enforce TypedPath's type parameters
+record TFileType<a> {
+  fileType: FileType,
+}
+
+// Rel Number represents number of directories up from base point
+enum Base {
+  Rel(Number),
+  Abs(AbsoluteRoot),
+},
+type PathInfo = (Base, FileType, List<String>),
+record TBase<a> {
+  base: Base,
+},
+type TypedPath<a, b> = (TBase<a>, TFileType<b>, List<String>),
+/**
+ * @section Types: Type declarations included in the Path module.
+ */
+
+/**
+ * Represents an absolute path's anchor point.
+ */
+export enum AbsoluteRoot {
+  Root,
+  Drive(Char),
+}
+
+/**
+ * Represents a system path.
+ */
+enum Path {
```

Comment:
```suggestion
export enum Path {
```
I think I want this exported... I want to be able to do something like:

```grain
match (Path.fromString("./my/relpath")) {
  AbsoluteFile(_) as path => path,
  AbsoluteDir(_) as path => Path.append(path, "index.md"),
  RelativeFile(_) as path => Path.append(Process.cwd(), path),
  RelativeDir(_) as path => Path.append(Path.append(Process.cwd(), path), "index.md")
}
```
and I don't see a (nice?) way to do that normalization if you make this opaque.

Reply:
Ok, I'll add `TypedPath`, `Relative`, etc into this section and write docblocks for them as well then